File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/2644-촌수 계산/solution.py
import sys
from collections import deque
si = sys.stdin.readline
n = int(si())
st, ed = list(map(int, si().split()))
m = int(si())

adj = [[] for _ in range(n + 1)]
dist = [-1] * (n + 1)

for _ in range(m):
    x, y = list(map(int, si().split()))
    adj[x].append(y)
    adj[y].append(x)

def bfs(x):
    queue = deque()
    queue.append(x)
    dist[x] = 0
    while queue:
        x = queue.popleft()
        for y in adj[x]:
            if dist[y] != -1:
                continue
            dist[y] = dist[x] + 1
            queue.append(y)

bfs(st)
print(dist[ed])

File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/2644-촌수 계산/solution.cpp
#include <iostream>
#include <algorithm>
#pragma warning(disable:4996)
using namespace std;
#define NM 105

#include <vector>
vector<int> adj[NM];
int n, m, st, ed, dist[NM];
void input() {
	cin >> n;
	cin >> st >> ed;
	cin >> m;
	for (int i = 1; i <= m; i++) {
		int u, v;
		cin >> u >> v;
		adj[u].push_back(v);
		adj[v].push_back(u);
	}
}

#include <queue>
void BFS(int S) {
	for (int i = 1; i <= n; i++) dist[i] = -1;
	queue<int> Q;
	Q.push(S), dist[S] = 0;
	while (!Q.empty()) {
		int x = Q.front(); Q.pop();
		for (int y : adj[x]) {
			if (dist[y] != -1) continue;
			dist[y] = dist[x] + 1;
			Q.push(y);
		}
	}
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	input();
	BFS(st);
	cout << dist[ed];
	return 0;
}